drivers/display/display_ili9340.c

```diff
@@ -185,120 +183,307 @@ static int ili9340_read(const struct device *dev, const u16_t x,
 		return 0;
+	} else if (pixel_format == PIXEL_FORMAT_RGB_565) {
+		arg = ILI9XXX_DATA_PIXEL_FORMAT_MCU_16_BIT |
+				ILI9XXX_DATA_PIXEL_FORMAT_RGB_16_BIT;
+	} else if (pixel_format == PIXEL_FORMAT_RGB_888) {
+		arg = ILI9XXX_DATA_PIXEL_FORMAT_MCU_18_BIT |
+				ILI9XXX_DATA_PIXEL_FORMAT_RGB_18_BIT;
+	} else {
+		LOG_ERR("Pixel format 0x%X is not supported", pixel_format);
+		return -ENOTSUP;
 	}
-	LOG_ERR("Pixel format change not implemented");
-	return -ENOTSUP;
+
+	rv = ili9xxx_transmit(data, ILI9XXX_CMD_PIXEL_FORMAT_SET, &arg, 1);
+	if (rv) {
+		LOG_ERR("Failed to change pix fmt %d -> %d",
+				data->pixel_format, pixel_format);
+		return rv;
+	}
+	data->pixel_format = pixel_format;
+	return 0;
 }
 
-static int ili9340_set_orientation(const struct device *dev,
+
+static int ili9xxx_set_orientation(const struct device *dev,
 				   const enum display_orientation orientation)
 {
-	if (orientation == DISPLAY_ORIENTATION_NORMAL) {
+	u8_t cmd;
+	struct ili9xxx_data *data = (struct ili9xxx_data *)dev->driver_data;
+
+	if (orientation == data->orientation) {
 		return 0;
 	}
-	LOG_ERR("Changing display orientation not implemented");
-	return -ENOTSUP;
+	/* todo: check if square */
+	switch (orientation) {
+	case DISPLAY_ORIENTATION_NORMAL:
+		cmd = 0;
+		break;
+	case DISPLAY_ORIENTATION_ROTATED_90:
+		cmd = ILI9XXX_DATA_MEM_ACCESS_CTRL_MV
+			| ILI9XXX_DATA_MEM_ACCESS_CTRL_MY;
+		break;
+	case DISPLAY_ORIENTATION_ROTATED_180:
+		cmd = ILI9XXX_DATA_MEM_ACCESS_CTRL_MX
+			| ILI9XXX_DATA_MEM_ACCESS_CTRL_MY;
+		break;
+	case DISPLAY_ORIENTATION_ROTATED_270:
+		cmd = ILI9XXX_DATA_MEM_ACCESS_CTRL_MV
+			| ILI9XXX_DATA_MEM_ACCESS_CTRL_MX;
+		break;
+	default:
+		return -EINVAL;
+	}
+#ifdef DT_INST_0_ILITEK_ILI9XXX_MADC
+	cmd |= DT_INST_0_ILITEK_ILI9XXX_MADC;
+#else
+	cmd |= ILI9XXX_DATA_MEM_ACCESS_CTRL_BGR;
+#endif
+	data->orientation = orientation;
+	return ili9xxx_transmit(data, ILI9XXX_CMD_MEM_ACCESS_CTRL, &cmd, sizeof(cmd));
 }
 
-static void ili9340_get_capabilities(const struct device *dev,
+static void ili9xxx_get_capabilities(const struct device *dev,
 				     struct display_capabilities *capabilities)
 {
+	const struct ili9xxx_data *data = (struct ili9xxx_data *)dev->driver_data;
 	memset(capabilities, 0, sizeof(struct display_capabilities));
-	capabilities->x_resolution = 320U;
-	capabilities->y_resolution = 240U;
-#ifdef CONFIG_ILI9340_RGB565
-	capabilities->supported_pixel_formats = PIXEL_FORMAT_RGB_565;
-	capabilities->current_pixel_format = PIXEL_FORMAT_RGB_565;
-#else
-	capabilities->supported_pixel_formats = PIXEL_FORMAT_RGB_888;
-	capabilities->current_pixel_format = PIXEL_FORMAT_RGB_888;
+	capabilities->x_resolution = RES_HOR;
+	capabilities->y_resolution = RES_VER;
+	capabilities->supported_pixel_formats =
+			PIXEL_FORMAT_RGB_565 | PIXEL_FORMAT_RGB_888;
+	capabilities->current_pixel_format = data->pixel_format;
+	capabilities->current_orientation = data->orientation;
+}
+
+#define EXEC_CMD(cmd) { \
```

Comment:
This would be hard to debug, return is opaque to the reader. `_` must not be used and anyway name should be ili93xx_exec_cmd(cmd).

Reply:
Done